### DISCUSSION TITLE
Block creation/injection API alias

### DISCUSSION THREAD

**Discussion by nexensys** [ScratchAddons/ScratchAddons#3967] [Help] [answered]
I have pretty much finished the block creation api. I currently have it set as the `injection` property of `addon.tab` I am thinking that it may be better to have it be `addon.tab.blocks`. Any opinions?

I was about to provide an example, but it stopped working... @GarboMuffin this is the topic, thanks for the help.

**Comment by nexensys** [comment]
I fixed it. Here is a feature preview:

```js
let api = addon.tab.injection;
let script = api.createBlock("event_whenflagclicked", {});
script
  .connect(api.createBlock("motion_movesteps", { inputs: { STEPS: 15 } }))
  .connect(api.createBlock("motion_turnright", { inputs: { DEGREES: 20 } }));
await script.toWorkspace();
```

Result:

![image](https://user-images.githubusercontent.com/70162741/147009722-aaf202d0-d748-42ec-866d-6878527da5d8.png)

Still need to fiddle with blocks that have stacks like `if`.

**Reply by GarboMuffin** [reply]
High level overview of my opinions of this API:

 - Trying to do a lot without a clear problem being solved
 - Bad performance (run that script in the player sprite of Appel...)
 - I don't like how it has a hardcoded list of every opcode, its inputs, and their types
 - Blockly already has an *interesting* and *unique* XML-based API for block-related functions
 - Having an easy way to programmatically create blocks is a great idea. But I don't think this is it.

Before you start making any changes and sinking more time into this, please wait until I've had more time to write some more complete thoughts (tomorrow, probably)

**Reply by nexensys** [reply]
Okay then...

**Reply by nexensys** [reply]
Thinking back...I really have not been very focused on efficiency. I was just looking through the vm and saw the `createBlock` function, then the only way to get the thing to show up was to refresh all of the blocks. Should probably use the Blockly dom api since I am already using it for procedures. I never noticed that issue because most of the scripts in my projects are on average nowhere near the size of @griffpatch's.

**Reply by GarboMuffin** [reply]
I was going to write a much longer comment but I'll keep it relatively short.

The largest problem with this API to me is that it's a solution looking for a problem and that it's scope is too large for an API with no proven use case.

Editing procedure prototypes is a reasonably small segment of code. You can just put that segment in the addon that needs it. If multiple addons need that segment then extract it to some shared module (which doesn't have to be in addon.tab). I also don't see why editing procedure prototypes and adding blocks need to be in the same API.

Actual addons are more important than large APIs with hypothetical use cases.

**Comment by apple502j** [comment, marked as answer]
Rejected. VM-based block manipulation works fine.
